django/utils/encoding.py

```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+# List of byte values that uri_to_iri is going to decode from percent encoding.
+# These are the unreserved characters from RFC 3986 and everything above 128,
+# because bytes >= 128 are part of multibyte unicode characters.
+_decode_ranges = [[45, 46, 95, 126], range(65, 91), range(97, 123), range(128, 256)]
+_hextobyte = {
+    (fmt % char).encode(): six.int2byte(char)
+    for range in _decode_ranges for char in range for fmt in ['%02x', '%02X']
+}
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
```

Comment:
IMHO, that is actually not that difficult. `2.2` and `4.1` are quite obvious and `6.1` references `RFC3491`, which just says:
>    This profile specifies prohibiting using the following tables from   [STRINGPREP]:
>
>   Table C.1.2
   Table C.2.2
   Table C.3
   Table C.4
   Table C.5
   Table C.6
   Table C.7
   Table C.8
   Table C.9

And it just so happens that `stringprep` is already part of the standardlib: https://docs.python.org/2/library/stringprep.html :).

The more relevant question is whether we actually want this. `uri_to_iri` is used to transform the page url for routing etc. This seems rather unwise to me, but I don't understand that infrastructure enough to want to touch it.

If we implement step 4, people need to understand step 4 in order to use django routing and make sense of `request.path`. Do we want that?